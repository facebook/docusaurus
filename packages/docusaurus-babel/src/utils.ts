/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {BABEL_CONFIG_FILE_NAME} from '@docusaurus/utils';
import type {TransformOptions} from '@babel/core';

export async function getCustomBabelConfigFilePath(
  siteDir: string,
): Promise<string | undefined> {
  const customBabelConfigurationPath = path.join(
    siteDir,
    BABEL_CONFIG_FILE_NAME,
  );
  return (await fs.pathExists(customBabelConfigurationPath))
    ? customBabelConfigurationPath
    : undefined;
}

export function getBabelOptions({
  isServer,
  babelOptions,
}: {
  isServer?: boolean;
  // TODO Docusaurus v4 fix this
  //  weird to have getBabelOptions take a babelOptions param
  babelOptions?: TransformOptions | string;
} = {}): TransformOptions {
  const caller = {name: isServer ? 'server' : 'client'};
  if (typeof babelOptions === 'string') {
    return {
      babelrc: false,
      configFile: babelOptions,
      caller,
    };
  }
  return {
    ...(babelOptions ?? {
      presets: [require.resolve('@docusaurus/babel/preset')],
    }),
    babelrc: false,
    configFile: false,
    caller,
  };
}
