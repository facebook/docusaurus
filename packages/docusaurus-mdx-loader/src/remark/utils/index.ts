/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import escapeHtml from 'escape-html';
import type {Parent, Node} from 'unist';
import type {PhrasingContent, Heading} from 'mdast';
import type {
  MdxJsxAttribute,
  MdxJsxAttributeValueExpression,
  MdxJsxTextElement,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';

/**
 * Util to transform one node type to another node type
 * The input node is mutated in place
 * @param node the node to mutate
 * @param newNode what the original node should become become
 */
export function transformNode<NewNode extends Node>(
  node: Node,
  newNode: NewNode,
): NewNode {
  Object.keys(node).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    delete node[key];
  });
  Object.keys(newNode).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    node[key] = newNode[key];
  });
  return node as NewNode;
}

export function stringifyContent(
  node: Parent,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  return (node.children as PhrasingContent[])
    .map((item) => toValue(item, toString))
    .join('');
}

// TODO This is really a workaround, and not super reliable
// For now we only support serializing tagName, className and content
// Can we implement the TOC with real JSX nodes instead of html strings later?
function mdxJsxTextElementToHtml(
  element: MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  const tag = element.name;

  const attributes = element.attributes.filter(
    (child): child is MdxJsxAttribute => child.type === 'mdxJsxAttribute',
  );

  const classAttribute =
    attributes.find((attr) => attr.name === 'className') ??
    attributes.find((attr) => attr.name === 'class');

  const classAttributeString = classAttribute
    ? `class="${escapeHtml(String(classAttribute.value))}"`
    : ``;

  const allAttributes = classAttributeString ? ` ${classAttributeString}` : '';

  const content = stringifyContent(element, toString);

  return `<${tag}${allAttributes}>${content}</${tag}>`;
}

export function toValue(
  node: PhrasingContent | Heading | MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  switch (node.type) {
    case 'mdxJsxTextElement': {
      return mdxJsxTextElementToHtml(node as MdxJsxTextElement, toString);
    }
    case 'text':
      return escapeHtml(node.value);
    case 'heading':
      return stringifyContent(node, toString);
    case 'inlineCode':
      return `<code>${escapeHtml(node.value)}</code>`;
    case 'emphasis':
      return `<em>${stringifyContent(node, toString)}</em>`;
    case 'strong':
      return `<strong>${stringifyContent(node, toString)}</strong>`;
    case 'delete':
      return `<del>${stringifyContent(node, toString)}</del>`;
    case 'link':
      return stringifyContent(node, toString);
    default:
      return toString(node);
  }
}

export function assetRequireAttributeValue(
  requireString: string,
  hash: string,
): MdxJsxAttributeValueExpression {
  return {
    type: 'mdxJsxAttributeValueExpression',
    value: `require("${requireString}").default${hash && ` + '${hash}'`}`,
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'BinaryExpression',
              left: {
                type: 'MemberExpression',
                object: {
                  type: 'CallExpression',
                  callee: {
                    type: 'Identifier',
                    name: 'require',
                  },
                  arguments: [
                    {
                      type: 'Literal',
                      value: requireString,
                      raw: `"${requireString}"`,
                    },
                  ],
                  optional: false,
                },
                property: {
                  type: 'Identifier',
                  name: 'default',
                },
                computed: false,
                optional: false,
              },
              operator: '+',
              right: {
                type: 'Literal',
                value: hash,
                raw: `"${hash}"`,
              },
            },
          },
        ],
        sourceType: 'module',
        comments: [],
      },
    },
  };
}
