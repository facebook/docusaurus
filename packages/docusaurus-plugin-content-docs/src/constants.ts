/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/** The name of the version that's actively worked on (e.g. `website/docs`) */
export const CURRENT_VERSION_NAME = 'current';
/** All doc versions are stored here by version names */
export const VERSIONED_DOCS_DIR = 'versioned_docs';
/** All doc versioned sidebars are stored here by version names */
export const VERSIONED_SIDEBARS_DIR = 'versioned_sidebars';
/** The version names. Should 1-1 map to the content of versioned docs dir. */
export const VERSIONS_JSON_FILE = 'versions.json';
