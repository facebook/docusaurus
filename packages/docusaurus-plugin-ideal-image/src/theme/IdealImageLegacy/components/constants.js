const load = 'load';
const loading = 'loading';
const loaded = 'loaded';
const error = 'error';
const noicon = 'noicon';
const offline = 'offline';

export const icons = {
  load,
  loading,
  loaded,
  error,
  noicon,
  offline,
};

const initial = 'initial';

export const loadStates = {
  initial,
  loading,
  loaded,
  error,
};
