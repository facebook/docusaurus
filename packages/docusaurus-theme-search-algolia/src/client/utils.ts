/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {FacetFilters} from 'algoliasearch/lite';

export function mergeFacetFilters(
  f1: FacetFilters,
  f2: FacetFilters,
): FacetFilters;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined {
  if (f1 === undefined) {
    return f2;
  }
  if (f2 === undefined) {
    return f1;
  }

  const normalize = (f: FacetFilters): FacetFilters =>
    typeof f === 'string' ? [f] : f;

  // Historical behavior: we flatten everything
  // TODO I'm pretty sure this is incorrect
  //  see https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/?client=javascript
  //  Note: Algolia is working to provide a reliable facet merging strategy
  //  see https://github.com/facebook/docusaurus/pull/11327#issuecomment-3284742923
  return [...normalize(f1), ...normalize(f2)];
}
