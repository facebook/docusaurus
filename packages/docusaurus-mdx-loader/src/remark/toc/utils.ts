/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {toValue} from '../utils';
import type {Node} from 'unist';
import type {
  MdxjsEsm,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';
import type {TOCHeading, TOCItem, TOCItems, TOCSlice} from './types';
import type {
  Program,
  SpreadElement,
  ImportDeclaration,
  ImportSpecifier,
} from 'estree';

export function getImportDeclarations(program: Program): ImportDeclaration[] {
  return program.body.filter(
    (item): item is ImportDeclaration => item.type === 'ImportDeclaration',
  );
}

export function isMarkdownImport(node: Node): node is ImportDeclaration {
  if (node.type !== 'ImportDeclaration') {
    return false;
  }
  const importPath = (node as ImportDeclaration).source.value;
  return typeof importPath === 'string' && /\.mdx?$/.test(importPath);
}

export function findDefaultImportName(
  importDeclaration: ImportDeclaration,
): string | undefined {
  return importDeclaration.specifiers.find(
    (o: Node) => o.type === 'ImportDefaultSpecifier',
  )?.local.name;
}

export function findNamedImportSpecifier(
  importDeclaration: ImportDeclaration,
  localName: string,
): ImportSpecifier | undefined {
  return importDeclaration?.specifiers.find(
    (specifier): specifier is ImportSpecifier =>
      specifier.type === 'ImportSpecifier' &&
      specifier.local.name === localName,
  );
}

// Before: import Partial from "partial"
// After: import Partial, {toc as __tocPartial} from "partial"
export function addTocSliceImportIfNeeded({
  importDeclaration,
  tocExportName,
  tocSliceImportName,
}: {
  importDeclaration: ImportDeclaration;
  tocExportName: string;
  tocSliceImportName: string;
}): void {
  // We only add the toc slice named import if it doesn't exist already
  if (!findNamedImportSpecifier(importDeclaration, tocSliceImportName)) {
    importDeclaration.specifiers.push({
      type: 'ImportSpecifier',
      imported: {type: 'Identifier', name: tocExportName},
      local: {type: 'Identifier', name: tocSliceImportName},
    });
  }
}

export function isNamedExport(
  node: Node,
  exportName: string,
): node is MdxjsEsm {
  if (node.type !== 'mdxjsEsm') {
    return false;
  }
  const program = (node as MdxjsEsm).data?.estree;
  if (!program) {
    return false;
  }
  if (program.body.length !== 1) {
    return false;
  }
  const exportDeclaration = program.body[0]!;
  if (exportDeclaration.type !== 'ExportNamedDeclaration') {
    return false;
  }
  const variableDeclaration = exportDeclaration.declaration;
  if (variableDeclaration?.type !== 'VariableDeclaration') {
    return false;
  }
  const {id} = variableDeclaration.declarations[0]!;
  if (id.type !== 'Identifier') {
    return false;
  }

  return id.name === exportName;
}

export async function createTOCExportNodeAST({
  tocExportName,
  tocItems,
}: {
  tocExportName: string;
  tocItems: TOCItems;
}): Promise<MdxjsEsm> {
  function createTOCSliceAST(tocSlice: TOCSlice): SpreadElement {
    return {
      type: 'SpreadElement',
      argument: {type: 'Identifier', name: tocSlice.importName},
    };
  }

  async function createTOCHeadingAST({heading}: TOCHeading) {
    const {toString} = await import('mdast-util-to-string');
    const {valueToEstree} = await import('estree-util-value-to-estree');
    const value: TOCItem = {
      value: toValue(heading, toString),
      id: heading.data!.id!,
      level: heading.depth,
    };
    return valueToEstree(value);
  }

  async function createTOCItemAST(tocItem: TOCItems[number]) {
    switch (tocItem.type) {
      case 'slice':
        return createTOCSliceAST(tocItem);
      case 'heading':
        return createTOCHeadingAST(tocItem);
      default: {
        throw new Error(`unexpected toc item type`);
      }
    }
  }

  return {
    type: 'mdxjsEsm',
    value: '', // See https://github.com/facebook/docusaurus/pull/9684#discussion_r1457595181
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name: tocExportName,
                  },
                  init: {
                    type: 'ArrayExpression',
                    elements: await Promise.all(tocItems.map(createTOCItemAST)),
                  },
                },
              ],
              kind: 'const',
            },
            specifiers: [],
            source: null,
          },
        ],
        sourceType: 'module',
      },
    },
  };
}
