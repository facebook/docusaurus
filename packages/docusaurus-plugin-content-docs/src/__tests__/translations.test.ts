/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {updateTranslationFileMessages} from '@docusaurus/utils';
import {CURRENT_VERSION_NAME} from '../constants';
import {
  getLoadedContentTranslationFiles,
  translateLoadedContent,
} from '../translations';
import type {
  DocMetadata,
  LoadedContent,
  LoadedVersion,
} from '@docusaurus/plugin-content-docs';

function createSampleDoc(doc: Pick<DocMetadata, 'id'>): DocMetadata {
  return {
    sourceDirName: '',
    draft: false,
    tags: [],
    editUrl: 'any',
    lastUpdatedAt: 0,
    lastUpdatedBy: 'any',
    next: undefined,
    previous: undefined,
    permalink: 'any',
    slug: 'any',
    source: 'any',
    version: 'any',
    title: `${doc.id} title`,
    frontMatter: {
      sidebar_label: `${doc.id} title`,
    },
    description: `${doc.id} description`,
    ...doc,
  };
}

function createSampleVersion(
  version: Pick<LoadedVersion, 'versionName'>,
): LoadedVersion {
  return {
    label: `${version.versionName} label`,
    path: '/docs/',
    routePriority: undefined,
    sidebarFilePath: 'any',
    isLast: true,
    contentPath: 'any',
    contentPathLocalized: 'any',
    tagsPath: '/tags/',
    banner: null,
    badge: true,
    className: '',
    drafts: [],
    docs: [
      createSampleDoc({id: 'doc1'}),
      createSampleDoc({id: 'doc2'}),
      createSampleDoc({id: 'doc3'}),
      createSampleDoc({id: 'doc4'}),
      createSampleDoc({id: 'doc5'}),
    ],
    sidebars: {
      docs: [
        {
          type: 'category',
          label: 'Getting started',
          collapsed: false,
          collapsible: true,
          link: {
            type: 'generated-index',
            slug: '/category/getting-started-index-slug',
            permalink: '/docs/category/getting-started-index-slug',
            title: 'Getting started index title',
            description: 'Getting started index description',
          },
          items: [
            {
              type: 'doc',
              id: 'doc1',
            },
            {
              type: 'doc',
              id: 'doc2',
              label: 'Second doc translatable',
              translatable: true,
            },
            {
              type: 'link',
              label: 'Link label',
              href: 'https://facebook.com',
            },
            {
              type: 'ref',
              id: 'doc1',
            },
          ],
        },
        {
          type: 'doc',
          id: 'doc3',
        },
      ],
      otherSidebar: [
        {
          type: 'doc',
          id: 'doc4',
        },
        {
          type: 'ref',
          id: 'doc5',
          label: 'Fifth doc translatable',
          translatable: true,
        },
      ],
    },
    ...version,
  };
}

const SampleLoadedContent: LoadedContent = {
  loadedVersions: [
    createSampleVersion({
      versionName: CURRENT_VERSION_NAME,
    }),
    createSampleVersion({
      versionName: '2.0.0',
    }),
    createSampleVersion({
      versionName: '1.0.0',
    }),
  ],
};

function getSampleTranslationFiles() {
  return getLoadedContentTranslationFiles(SampleLoadedContent);
}
function getSampleTranslationFilesTranslated() {
  const translationFiles = getSampleTranslationFiles();
  return translationFiles.map((translationFile) =>
    updateTranslationFileMessages(
      translationFile,
      (message) => `${message} (translated)`,
    ),
  );
}

describe('getLoadedContentTranslationFiles', () => {
  it('returns translation files', () => {
    expect(getSampleTranslationFiles()).toMatchSnapshot();
  });
});

describe('translateLoadedContent', () => {
  it('does not translate anything if translation files are untranslated', () => {
    const translationFiles = getSampleTranslationFiles();
    expect(
      translateLoadedContent(SampleLoadedContent, translationFiles),
    ).toEqual(SampleLoadedContent);
  });

  it('returns translated loaded content', () => {
    const translationFiles = getSampleTranslationFilesTranslated();
    expect(
      translateLoadedContent(SampleLoadedContent, translationFiles),
    ).toMatchSnapshot();
  });
});
