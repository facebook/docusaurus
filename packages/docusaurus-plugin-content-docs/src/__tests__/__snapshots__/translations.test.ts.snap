// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLoadedContentTranslationFiles returns translation files 1`] = `
[
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.doc.Second doc translatable": {
        "description": "The label for the doc item Second doc translatable in sidebar docs, linking to the doc doc2",
        "message": "Second doc translatable",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "sidebar.otherSidebar.doc.Fifth doc translatable": {
        "description": "The label for the doc item Fifth doc translatable in sidebar otherSidebar, linking to the doc doc5",
        "message": "Fifth doc translatable",
      },
      "version.label": {
        "description": "The label for version current",
        "message": "current label",
      },
    },
    "path": "current",
  },
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.doc.Second doc translatable": {
        "description": "The label for the doc item Second doc translatable in sidebar docs, linking to the doc doc2",
        "message": "Second doc translatable",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "sidebar.otherSidebar.doc.Fifth doc translatable": {
        "description": "The label for the doc item Fifth doc translatable in sidebar otherSidebar, linking to the doc doc5",
        "message": "Fifth doc translatable",
      },
      "version.label": {
        "description": "The label for version 2.0.0",
        "message": "2.0.0 label",
      },
    },
    "path": "version-2.0.0",
  },
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.doc.Second doc translatable": {
        "description": "The label for the doc item Second doc translatable in sidebar docs, linking to the doc doc2",
        "message": "Second doc translatable",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "sidebar.otherSidebar.doc.Fifth doc translatable": {
        "description": "The label for the doc item Fifth doc translatable in sidebar otherSidebar, linking to the doc doc5",
        "message": "Fifth doc translatable",
      },
      "version.label": {
        "description": "The label for version 1.0.0",
        "message": "1.0.0 label",
      },
    },
    "path": "version-1.0.0",
  },
]
`;

exports[`translateLoadedContent returns translated loaded content 1`] = `
{
  "loadedVersions": [
    {
      "badge": true,
      "banner": null,
      "className": "",
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc1 title",
          },
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc1 title",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc2 title",
          },
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc2 title",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc3 title",
          },
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc3 title",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc4 title",
          },
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc4 title",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc5 title",
          },
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc5 title",
          "version": "any",
        },
      ],
      "drafts": [],
      "isLast": true,
      "label": "current label (translated)",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "collapsible": true,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "label": "Second doc translatable (translated)",
                "translatable": true,
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "label": "Fifth doc translatable (translated)",
            "translatable": true,
            "type": "ref",
          },
        ],
      },
      "tagsPath": "/tags/",
      "versionName": "current",
    },
    {
      "badge": true,
      "banner": null,
      "className": "",
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc1 title",
          },
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc1 title",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc2 title",
          },
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc2 title",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc3 title",
          },
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc3 title",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc4 title",
          },
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc4 title",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc5 title",
          },
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc5 title",
          "version": "any",
        },
      ],
      "drafts": [],
      "isLast": true,
      "label": "2.0.0 label (translated)",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "collapsible": true,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "label": "Second doc translatable (translated)",
                "translatable": true,
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "label": "Fifth doc translatable (translated)",
            "translatable": true,
            "type": "ref",
          },
        ],
      },
      "tagsPath": "/tags/",
      "versionName": "2.0.0",
    },
    {
      "badge": true,
      "banner": null,
      "className": "",
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc1 title",
          },
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc1 title",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc2 title",
          },
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc2 title",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc3 title",
          },
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc3 title",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc4 title",
          },
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc4 title",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "draft": false,
          "editUrl": "any",
          "frontMatter": {
            "sidebar_label": "doc5 title",
          },
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "slug": "any",
          "source": "any",
          "sourceDirName": "",
          "tags": [],
          "title": "doc5 title",
          "version": "any",
        },
      ],
      "drafts": [],
      "isLast": true,
      "label": "1.0.0 label (translated)",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "collapsible": true,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "label": "Second doc translatable (translated)",
                "translatable": true,
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "label": "Fifth doc translatable (translated)",
            "translatable": true,
            "type": "ref",
          },
        ],
      },
      "tagsPath": "/tags/",
      "versionName": "1.0.0",
    },
  ],
}
`;
